// @remotion/skia/dist/esm/enable.mjs
import { createRequire } from "node:module";
var __require = /* @__PURE__ */ createRequire(import.meta.url);

// src/enable.ts
import fs from "fs";
import { webpack } from "@remotion/bundler";
var enableSkia = (currentConfiguration) => {
  const newExtensions = [
    ".web.js",
    ".web.ts",
    ".web.tsx",
    ...currentConfiguration.resolve?.extensions ?? []
  ];
  const deduplicatedExtensions = [...new Set(newExtensions)];
  return {
    ...currentConfiguration,
    plugins: [
      ...currentConfiguration.plugins ?? [],
      new class CopySkiaPlugin {
        apply(compiler) {
          compiler.hooks.thisCompilation.tap("AddSkiaPlugin", (compilation) => {
            compilation.hooks.processAssets.tapPromise({
              name: "copy-skia",
              stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL
            }, async () => {
              const src = __require.resolve("canvaskit-wasm/bin/full/canvaskit.wasm");
              if (compilation.getAsset(src)) {
                return;
              }
              compilation.emitAsset("/canvaskit.wasm", new webpack.sources.RawSource(await fs.promises.readFile(src)));
            });
          });
        }
      }
    ],
    resolve: {
      ...currentConfiguration.resolve,
      fallback: {
        fs: false,
        path: false
      },
      extensions: deduplicatedExtensions,
      alias: {
        ...currentConfiguration.resolve?.alias,
        "react-native-reanimated": "require('react-native-reanimated')",
        "react-native-reanimated/lib/reanimated2/core": "require('react-native-reanimated/lib/reanimated2/core')",
        "react-native/Libraries/Image/AssetRegistry": false
      }
    }
  };
};
export {
  enableSkia
};
