// @remotion/gif/dist/cjs/Gif.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Gif = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const remotion_1 = require("remotion");
const GifForDevelopment_1 = require("./GifForDevelopment");
const GifForRendering_1 = require("./GifForRendering");
/*
 * @description Displays a GIF that synchronizes with Remotions useCurrentFrame().
 * @see [Documentation](https://remotion.dev/docs/gif)
 */
const gifSchema = {
    playbackRate: {
        type: 'number',
        min: 0,
        max: 10,
        step: 0.1,
        default: 1,
        description: 'Playback Rate',
    },
    'style.translate': {
        type: 'translate',
        step: 1,
        default: '0px 0px',
        description: 'Position',
    },
    'style.scale': {
        type: 'number',
        min: 0.05,
        max: 100,
        step: 0.01,
        default: 1,
        description: 'Scale',
    },
    'style.rotate': {
        type: 'rotation',
        step: 1,
        default: '0deg',
        description: 'Rotation',
    },
    'style.opacity': {
        type: 'number',
        min: 0,
        max: 1,
        step: 0.01,
        default: 1,
        description: 'Opacity',
    },
};
const GifInner = ({ src, width, height, onLoad, onError, fit, playbackRate, loopBehavior, id, delayRenderTimeoutInMilliseconds, durationInFrames, style, controls, ref, ...sequenceProps }) => {
    const env = (0, remotion_1.useRemotionEnvironment)();
    const { durationInFrames: videoDuration } = (0, remotion_1.useVideoConfig)();
    const resolvedDuration = durationInFrames !== null && durationInFrames !== void 0 ? durationInFrames : videoDuration;
    const gifProps = {
        src,
        width,
        height,
        onLoad,
        onError,
        fit,
        playbackRate,
        loopBehavior,
        id,
        delayRenderTimeoutInMilliseconds,
        style,
    };
    const inner = env.isRendering ? (jsx_runtime_1.jsx(GifForRendering_1.GifForRendering, { ...gifProps, ref: ref })) : (jsx_runtime_1.jsx(GifForDevelopment_1.GifForDevelopment, { ...gifProps, ref: ref }));
    return (jsx_runtime_1.jsx(remotion_1.Sequence, { layout: "none", durationInFrames: resolvedDuration, name: "<Gif>", controls: controls, ...sequenceProps, children: inner }));
};
exports.Gif = remotion_1.Internals.wrapInSchema(GifInner, gifSchema);
exports.Gif.displayName = 'Gif';
remotion_1.Internals.addSequenceStackTraces(exports.Gif);
